### PR TITLE
Added method to disable all non-core Proposals

### PR DESCRIPTION
**Pull request by rhafer**
This is needed to keep the non-core Cookbooks from being executed when a
node is coming back to "ready" after performing and OS upgrade. It's supposed
to be called before running the actual upgrade (i.e. before calling
prepare_nodes_for_os_upgrade()). Unfortunately we can't use
ServiceObject.destroy_active() for this, because that would cause (queued)
chef-client runs on the non-upgrades nodes. We don't want that here.

### CODE REVIEW

File: crowbar_framework/app/models/crowbar_service.rb [hunk 1]
```diff
@@ -176,6 +176,27 @@ def create_proposal
     base
   end
 
+  def disable_non_core_proposals
```
Comment (rhafer) [inline crowbar_framework/app/models/crowbar_service.rb]: I am not completely sure if CrowbarService is really the right place for this. I am interested in feedback.

**Comment by vuntz** [comment]
> This is needed to keep the non-core Cookbooks from being executed when a
> node is coming back to "ready" after performing and OS upgrade.

Oh, hrm.

Don't we need to also mark the proposals as not applied anymore? In which case maybe we need to combine both actions somehow (and do some apply_role to do the magic, but without running chef-client). Just an idea, might be more complicated than your approach.

**Comment by rhafer** [comment]
> Don't we need to also mark the proposals as not applied anymore? 

Not sure what you mean by this, but when I delete in the ```*-config-*``` roles (that's what the ```role.destroy``` the proposals show up as not being applied in the UI. Do you mean something else?

**Comment by vuntz** [comment]
@rhafer I meant exactly that. I just didn't read that closely enough.

So... I would possibly put this in a module in `lib/`... But @tboerger @rsalevsky @MaximilianMeister know better than me.

File: crowbar_framework/app/models/crowbar_service.rb (hunk 1 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/models/crowbar_service.rb]: As we have the architecture now with CrowbarService overriding/extending the ServiceObject I think it makes sense here.
Not sure if we manipulate much on NodeObjects in the lib either.

**Comment by rhafer** [comment]
This should be ready for more review now. I opted for keeping the code in crowbar_service, because @MaximilianMeister is ok with it. And because that we're currently have almost all of the upgrade related calls.

**Comment by MaximilianMeister** [comment]
codewise +1

File: crowbar_framework/app/models/crowbar_service.rb [hunk 2]
```diff
@@ -262,6 +262,27 @@ def create_proposal
...
+      @logger.debug("deactivating proposal #{role.name}")
+
+      roles_to_remove = role.elements.keys
+      roles_to_remove << role.name
+
+      roles_to_remove.each do |delete_role|
+        nodes = NodeObject.find("roles:#{delete_role}")
+        nodes.each do |node|
+          node.delete_from_run_list(delete_role)
+          node.save
```
Comment (jsuchome) [inline crowbar_framework/app/models/crowbar_service.rb]: Just thinking: aren't we saving node(s) quite often this way? Wouldn't it be better to iterate over all nodes, for each one delete all roles that need to be removed and than save the node only once?

(But maybe it's not relevant or node.save is so fast that we do not need to bother...)

**Comment by jsuchome** [comment]
This looks good to me. +1 (but not tested)

File: crowbar_framework/app/models/crowbar_service.rb (hunk 2 above)
Comment (rhafer) [inline crowbar_framework/app/models/crowbar_service.rb]: Hm, you're right. I'll add that as a followup item to trello. But I'll merge this variant for now.

**Pull request merged**